qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -266,63 +251,182 @@ def run(self, dag):
+            mapped_dag = dag.copy_empty_like()
+            _apply_sabre_result(
+                mapped_dag,
+                dag,
+                self._qubit_indices,
+                canonical_register,
+                original_layout,
+                sabre_result,
+            )
+            return mapped_dag
+        return dag
+
+
+def _build_sabre_dag(dag, num_qubits, num_clbits, qubit_indices, clbit_indices):
+    dag_list = []
+    node_blocks = {}
+    for node in dag.topological_op_nodes():
+        cargs = {clbit_indices[x] for x in node.cargs}
+        if node.op.condition is not None:
+            for clbit in dag._bits_in_condition(node.op.condition):
+                cargs.add(clbit_indices[clbit])
+        if isinstance(node.op, ControlFlowOp):
+            node_blocks[node._node_id] = [
+                # TODO: what's the easiest way to get the smallest num_qubits we
+                #  can pass here? A block dag should have fewer qubits than the
+                #  parent, so we could just use that, but really we only need
+                #  max(qubits) + 1
+                _build_sabre_dag(
+                    circuit_to_dag(block),
+                    num_qubits,
+                    num_clbits,
+                    {inner: qubit_indices[outer] for inner, outer in zip(block.qubits, node.qargs)},
+                    clbit_indices,
                 )
-                apply_gate(
-                    mapped_dag,
-                    node,
-                    original_layout,
+                for block in node.op.blocks
+            ]
+        dag_list.append(
+            (
+                node._node_id,
+                [qubit_indices[x] for x in node.qargs],
+                cargs,
+            )
+        )
+    return SabreDAG(num_qubits, num_clbits, dag_list, node_blocks)
+
+
+def _apply_sabre_result(
+    mapped_dag,
+    root_dag,
+    qubit_indices,
+    canonical_register,
+    initial_layout,
+    sabre_result,
+    component_map=None,
+):
+    bit_to_qreg_idx = {bit: idx for idx, bit in enumerate(canonical_register)}
+
+    def empty_dag(node):
+        out = DAGCircuit()
+        for qreg in mapped_dag.qregs.values():
+            out.add_qreg(qreg)
+        out.add_clbits(node.cargs)
+        return out
+
+    def apply_inner(out_dag, current_layout, qubit_indices_inner, result, id_to_node):
+        for node_id in result.node_order:
+            node = id_to_node[node_id]
+            if isinstance(node.op, ControlFlowOp):
+                # Handle control flow op and continue.
+                block_results = result.node_block_results[node_id]
+                mapped_block_dags = []
+                idle_qubits = set(out_dag.qubits)
+                for block, block_result in zip(node.op.blocks, block_results):
+                    # TODO: cache DAGs using id(block) as key when building sabre dag?
+                    block_id_to_node = circuit_to_dag(block)._multi_graph
```

Comment:
We can do this in a follow up, if there is insufficient time to add this before merging. I'd actually be curious to see how much of a runtime impact removing one round of circuit->dag conversions has. In general it should be super quick, but it does scale linearly with the number of instructions in the blocks.

Reply:
Done in 88d69e2. I didn't bother to benchmark, but it's an easy enough optimization.